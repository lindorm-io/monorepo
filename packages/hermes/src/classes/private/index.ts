export * from "./HermesMetadata";
export * from "./HermesRegistry";
export * from "./HermesScanner";
