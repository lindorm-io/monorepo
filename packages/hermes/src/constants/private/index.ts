export * from "./checksum-store";
export * from "./domain";
export * from "./event-store";
export * from "./saga-store";
export * from "./view-store";
