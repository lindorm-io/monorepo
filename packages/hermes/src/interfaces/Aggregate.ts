import { ClassLike, Dict } from "@lindorm/types";
import { AggregateData } from "../types";
import { IHermesMessage } from "./HermesMessage";

export interface IAggregate<S extends Dict = Dict> extends AggregateData {
  apply(causation: IHermesMessage, event: ClassLike): Promise<void>;
  load(event: IHermesMessage): void;
  toJSON(): AggregateData<S>;
}
