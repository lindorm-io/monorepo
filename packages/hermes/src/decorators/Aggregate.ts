import { Constructor } from "@lindorm/types";
import { AggregateDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Aggregate(options: AggregateDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addAggregate({
      encryption: options.encryption || false,
      name: options.name || name,
      namespace: options.namespace || null,
      target: target as unknown as Constructor,
    });
  };
}
