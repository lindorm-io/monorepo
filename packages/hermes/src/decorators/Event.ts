import { Constructor } from "@lindorm/types";
import { EventDecoratorOptions } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function Event(options?: EventDecoratorOptions): ClassDecorator {
  return function (target) {
    const { name, version } = extractNameData(target.name);
    globalHermesMetadata.addEvent({
      name: options?.name || name,
      target: target as unknown as Constructor,
      version: options?.version || version,
    });
  };
}
