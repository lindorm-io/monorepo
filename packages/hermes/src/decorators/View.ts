/* eslint-disable @typescript-eslint/no-unsafe-function-type */

import { isArray } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { ViewDecoratorOptions, ViewStoreSource } from "../types";
import { extractNameData, globalHermesMetadata } from "../utils/private";

export function View(
  AggregateClass: Constructor | Array<Constructor>,
  source: ViewStoreSource,
  options: ViewDecoratorOptions = {},
): ClassDecorator {
  return function (target) {
    const { name } = extractNameData(target.name);
    globalHermesMetadata.addView({
      aggregates: isArray(AggregateClass) ? AggregateClass : [AggregateClass],
      name: options.name || name,
      namespace: options.namespace || null,
      source,
      target: target as unknown as Constructor,
    });
  };
}
