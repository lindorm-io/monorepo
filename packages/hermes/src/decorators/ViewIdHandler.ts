import { Constructor } from "@lindorm/types";
import { ViewIdHandlerDescriptor } from "../types";
import { globalHermesMetadata } from "../utils/private";

export function ViewIdHandler<C extends Constructor>(
  EventClass: C,
): ViewIdHandlerDescriptor<C> {
  return function (target, key, descriptor) {
    globalHermesMetadata.addHandler({
      conditions: null,
      decorator: "ViewIdHandler",
      encryption: false,
      handler: descriptor.value,
      key: key.toString(),
      schema: null,
      target: target.constructor as Constructor,
      trigger: EventClass,
    });
  };
}
