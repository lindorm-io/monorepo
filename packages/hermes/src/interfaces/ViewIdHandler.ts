import { Constructor } from "@lindorm/types";
import { HandlerIdentifier, ViewIdCallback } from "../types";
import { NameData } from "../utils/private";

export interface IViewIdHandler<C extends Constructor = Constructor> {
  aggregate: HandlerIdentifier;
  event: NameData;
  key: string;
  view: HandlerIdentifier;
  handler: ViewIdCallback<C>;
}
