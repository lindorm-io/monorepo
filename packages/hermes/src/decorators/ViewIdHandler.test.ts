import { ViewIdCtx } from "../types";
import { globalHermesMetadata } from "../utils/private";
import { Aggregate } from "./Aggregate";
import { Event } from "./Event";
import { View } from "./View";
import { ViewIdHandler } from "./ViewIdHandler";

describe("ViewIdHandler Decorator", () => {
  @Aggregate()
  class TestViewAggregate {}

  test("should add metadata", () => {
    @Event()
    class TestViewIdentityHandlerEvent {
      public constructor(public readonly one: string) {}
    }

    @View(TestViewAggregate, "mongo")
    class TestViewIdentityHandlerView {
      @ViewIdHandler(TestViewIdentityHandlerEvent)
      public getIdTestViewIdentityHandlerEvent(
        ctx: ViewIdCtx<TestViewIdentityHandlerEvent>,
      ) {
        return ctx.event.one;
      }
    }

    expect(globalHermesMetadata.getView(TestViewIdentityHandlerView)).toMatchSnapshot();
  });
});
