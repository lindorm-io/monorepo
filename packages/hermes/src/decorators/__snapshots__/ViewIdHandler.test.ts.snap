// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewIdHandler Decorator should add metadata 1`] = `
{
  "aggregates": [
    [Function],
  ],
  "handlers": [
    {
      "conditions": null,
      "decorator": "ViewIdHandler",
      "encryption": false,
      "handler": [Function],
      "key": "getIdTestViewIdentityHandlerEvent",
      "schema": null,
      "target": [Function],
      "trigger": [Function],
    },
  ],
  "name": "test_view_identity_handler_view",
  "namespace": null,
  "source": "mongo",
  "target": [Function],
}
`;
